### DISCUSSION TITLE
Best way to host "secret" streams?

### DISCUSSION THREAD

**Discussion by cbix** [owncast/owncast#649] [Q&A]
We're evaluating Owncast to host, among other events, online movie nights for a group of friends. For this purpose it would be nice to have
- Viewer authentication,
- an access password for watching a stream or
- an easy way to generate a random/unpredictable stream URL

What's the best way to do this?

Thanks for any hints :)

Update: Did some testing now and found [livego](https://github.com/gwuhaolin/livego), which is a really simple RTMP server. What's nice is it lets me "create a stream" through the API on demand, so I'll just need to write a simple web frontend for admin and viewer.

**Comment by gabek** [comment]
I think your best options live within a proxy in front of Owncast.  @aaronpk mentioned he is requiring authentication for his streams using nginx here https://github.com/owncast/owncast/issues/646#issuecomment-763576630.

Other ideas, if you're inviting just a few friends to a stream to watch a movie, maybe think of whitelisting just those IPs in your proxy.  That would keep others out and really lock you down.  As for the URL, you can point any hostname you want to your Owncast instance, or no name at all.  Some people are running Owncast without DNS.

**Reply by cbix** [reply]
Thanks for the reference! I was already thinking about using HTTP basic auth like that if there's no builtin solution.

Wouldn't it be quite simple to implement a (client side) button to generate a random stream URL through the API?

**Reply by gabek** [reply]
I guess I don't understand what you mean by a random stream URL.  You control your own DNS, so you can make your hostname as random as you like.

**Reply by cbix** [reply]
Not the hostname, but the URL path. So instead of https://my.owncast/live, the stream would be accessible under https://my.owncast/lidooxeiBohphe0e. This is common practice for video conferences (instead of using a password) because it's a good compromise of user-friendliness and access security.

**Reply by gabek** [reply]
If you're just looking to have the web interface at a different URL you can embed the player into any page you like, at any url you like, behind any authentication or security you like.

There's discussion over at https://github.com/owncast/owncast/pull/631 about supporting actually hosting the entire web app from different paths, but again, that's not something Owncast itself is doing, it will always serve itself from root.  Serving from a different directory, especially a random one, would be a pretty niche need.

**Reply by aaronpk** [reply]
I think they're talking about a feature more like YouTube where each stream has a different URL, and "unlisted" streams aren't advertised on the home page at all. It's not serving the whole app from a different path, just the video stream itself. Kind of a big change compared to how OwnCast works right now but might be worth a discussion, especially in regards to recording streams too.

**Reply by cbix** [reply]
@aaronpk right, I think my misconception was that I didn't know Owncast is intended to host only a single stream per server! Of course in that case it would involve bigger changes to Owncast itself... Thanks for the explanation :)

I'll do a test setup now and maybe come back with a feature request if I get a more specific idea how this could work.

**Comment by gabek** [comment]
Just wanted a drop a mention of OAuth2 Proxy.  While I haven't tried it personally, it seems like a solution to throw in front of your Owncast server and supports a bunch of OAuth providers out of the box.  As a bonus it even works as a SSL termination solution.

https://oauth2-proxy.github.io/

**Reply by aaronpk** [reply]
Yep Vouch has a similar feature set which is the one I've been using https://github.com/vouch/vouch-proxy

**Reply by gabek** [reply]
What are you using as your OAuth provider with your streams that are authenticated?  Just curious as maybe we can write up some documentation pointing to a setup that works for people, and can make some suggestions for those looking to do similarly.

**Reply by geekgonecrazy** [reply]
Yup personally super interested in your setup @aaronpk I assume the idea would be a restricted user list so not something like Google
